tests/beta/interfaces/test_tape_tf.py

```diff
@@ -0,0 +1,741 @@
+
+        jac = tape.jacobian(res2, [a, b])
+        expected = [
+            [-2 * tf.sin(2 * a), 2 * tf.sin(2 * a) * tf.sin(b)],
+            [0, -tf.cos(2 * a) * tf.cos(b)],
+        ]
+        assert np.allclose(jac, expected, atol=tol, rtol=0)
+
+    def test_classical_processing(self, tol):
+        """Test classical processing within the quantum tape"""
+        a = tf.Variable(0.1, dtype=tf.float64)
+        b = tf.constant(0.2, dtype=tf.float64)
+        c = tf.Variable(0.3, dtype=tf.float64)
+
+        dev = qml.device("default.qubit", wires=1)
+
+        with tf.GradientTape() as tape:
+            with TFInterface.apply(QuantumTape()) as qtape:
+                qml.RY(a * c, wires=0)
+                qml.RZ(b, wires=0)
+                qml.RX(c + c ** 2 + tf.sin(a), wires=0)
+                expval(qml.PauliZ(0))
+
+            assert qtape.trainable_params == {0, 2}
+            assert qtape.get_parameters() == [a * c, c + c ** 2 + tf.sin(a)]
+            res = qtape.execute(dev)
+
+        res = tape.jacobian(res, [a, b, c])
+        assert isinstance(res[0], tf.Tensor)
+        assert res[1] is None
+        assert isinstance(res[2], tf.Tensor)
+
+    def test_no_trainable_parameters(self, tol):
+        """Test evaluation and Jacobian if there are no trainable parameters"""
+        dev = qml.device("default.qubit", wires=2)
+
+        with tf.GradientTape() as tape:
+
+            with TFInterface.apply(QuantumTape()) as qtape:
+                qml.RY(0.2, wires=0)
+                qml.RX(tf.constant(0.1), wires=0)
+                qml.CNOT(wires=[0, 1])
+                expval(qml.PauliZ(0))
+                expval(qml.PauliZ(1))
+
+            assert qtape.trainable_params == set()
+
+            res = qtape.execute(dev)
+
+        assert res.shape == (2,)
+        assert isinstance(res, tf.Tensor)
+
+    @pytest.mark.parametrize("U", [tf.constant([[0, 1], [1, 0]]), np.array([[0, 1], [1, 0]])])
+    def test_matrix_parameter(self, U, tol):
+        """Test that the TF interface works correctly
+        with a matrix parameter"""
+        a = tf.Variable(0.1, dtype=tf.float64)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        with tf.GradientTape() as tape:
+
+            with TFInterface.apply(QuantumTape()) as qtape:
+                qml.QubitUnitary(U, wires=0)
+                qml.RY(a, wires=0)
+                expval(qml.PauliZ(0))
+
+            assert qtape.trainable_params == {1}
+            res = qtape.execute(dev)
+
+        assert np.allclose(res, -tf.cos(a), atol=tol, rtol=0)
+
+        res = tape.jacobian(res, a)
+        assert np.allclose(res, tf.sin(a), atol=tol, rtol=0)
+
+    def test_differentiable_expand(self, mocker, tol):
+        """Test that operation and nested tapes expansion
+        is differentiable"""
+        mock = mocker.patch.object(qml.operation.Operation, "do_check_domain", False)
+
+        class U3(qml.U3):
+            def expand(self):

```

Comment:
Just for my own understanding: would the expansion logic supersede decompositions in the bigger picture?
```python
                tape._ops += [
                    qml.Rot(lam, theta, -lam, wires=wires),
                    qml.PhaseShift(phi + lam, wires=wires),
                ]
```

Reply:
Yep, exactly! Expansion is the replacement approach for decompositions :)